### PR TITLE
arch/risc-v: add support for capture driver on ESP32C6 and ESP32H2.

### PR DESCRIPTION
**Pull request by fdcavalcanti**
## Summary
This PR adds support for the capture driver on RISC-V ESP32C6 and ESP32H2. It is based on the MCPWM peripheral.
The basic functionality of MCPWM capture is to record the time when any pulse edge of the capture signal turns active.

Implementation comments:

- Having only one measurement is not sufficient to estimate frequency or duty cycle, because you still need at least one repetition of the pulse. In this case, 4 pulses are expected before a calculation is made.
- Up to 3 channels are possible (/dev/capture0, 1 and 2) and they are independent.

## Impact
Allows use of the capture driver.

## Testing
Tested on ESP32C6, ESP32H2 with the capture example using two ESP boards.
Also tested from 20 to 2000 Hz with varying duty cycle.

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/risc-v/src/common/espressif/esp_mcpwm.c [hunk 1]
```diff
@@ -0,0 +1,682 @@
...
+#include "esp_clk_tree.h"
+
+#ifdef CONFIG_ESP_MCPWM
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifdef CONFIG_ARCH_CHIP_ESP32H2
+#define MCPWM_DEV_CLK_SOURCE MCPWM_TIMER_CLK_SRC_PLL96M
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/espressif/esp_mcpwm.c:56]:
```suggestion
#  define MCPWM_DEV_CLK_SOURCE MCPWM_TIMER_CLK_SRC_PLL96M
```

File: arch/risc-v/src/common/espressif/esp_mcpwm.c [hunk 2]
```diff
@@ -0,0 +1,682 @@
...
+#else 
+#define MCPWM_DEV_CLK_SOURCE SOC_MOD_CLK_PLL_F160M
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/espressif/esp_mcpwm.c:58]:
```suggestion
#  define MCPWM_DEV_CLK_SOURCE SOC_MOD_CLK_PLL_F160M
```

File: arch/risc-v/src/common/espressif/esp_mcpwm.c [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+#endif
+
+#define MCPWM_DEV_CLK_PRESCALE 4
+#define MCPWM_CAPTURE_DEFAULT_GROUP 0
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/espressif/esp_mcpwm.c:62]:
```suggestion
#define MCPWM_DEV_CLK_PRESCALE      4
#define MCPWM_CAPTURE_DEFAULT_GROUP 0
```

File: arch/risc-v/src/common/espressif/esp_mcpwm.c [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+  MCPWM_CAP_CHANNEL_MAX /* Number of MCPWM capture channels */
+};
+
+/* Lowe-half data structure for a capture channel */
+
+struct mcpwm_cap_channel_lowerhalf_s
+{
+  /* The following block is part of the upper-half device struct */
+
+  FAR const struct cap_ops_s *ops;
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/espressif/esp_mcpwm.c:102]:
```suggestion
  const struct cap_ops_s *ops;
```

File: arch/risc-v/src/common/espressif/esp_mcpwm.c [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+ *   Returns OK on success.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE
+static int esp_capture_start(struct cap_lowerhalf_s *lower)
+{
+  struct mcpwm_cap_channel_lowerhalf_s *priv = (
+    struct mcpwm_cap_channel_lowerhalf_s *)lower;
+  mcpwm_hal_context_t *hal = &priv->common->hal;
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/espressif/esp_mcpwm.c:231]: accessed before `DEBUGASSERT(priv != NULL);`

File: arch/risc-v/src/common/espressif/esp_mcpwm.c [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+
+  DEBUGASSERT(priv != NULL);
+  DEBUGASSERT(priv->common->initialized);
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/espressif/esp_mcpwm.c:234]: is `DEBUGASSERT(priv->common != NULL);` needed?

File: arch/risc-v/src/common/espressif/esp_mcpwm.c [hunk 7]
```diff
@@ -0,0 +1,682 @@
...
+ *   Returns OK on success.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE
+static int esp_capture_stop(struct cap_lowerhalf_s *lower)
+{
+  struct mcpwm_cap_channel_lowerhalf_s *priv = (
+    struct mcpwm_cap_channel_lowerhalf_s *)lower;
+  mcpwm_hal_context_t *hal = &priv->common->hal;
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/espressif/esp_mcpwm.c:280]: Do we need `DEBUGASSERT(priv != NULL);` and similar checks like in `esp_capture_start`?

File: arch/risc-v/src/common/espressif/esp_mcpwm.c [hunk 8]
```diff
@@ -0,0 +1,682 @@
...
+ *   duty  - uint8_t pointer where the duty cycle value is written.
+ *
+ * Returned Value:
+ *   Returns OK on success.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE
+static int esp_capture_getduty(struct cap_lowerhalf_s *lower,
+                                 uint8_t *duty)
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/espressif/esp_mcpwm.c:313]:
```suggestion
static int esp_capture_getduty(struct cap_lowerhalf_s *lower,
                               uint8_t *duty)
```

File: arch/risc-v/src/common/espressif/esp_mcpwm.c [hunk 9]
```diff
@@ -0,0 +1,682 @@
...
+ *   duty  - uint8_t pointer where the frequency value is written.
+ *
+ * Returned Value:
+ *   Returns OK on success.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE
+static int esp_capture_getfreq(struct cap_lowerhalf_s *lower,
+                                 uint32_t *freq)
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/espressif/esp_mcpwm.c:341]:
```suggestion
static int esp_capture_getfreq(struct cap_lowerhalf_s *lower,
                               uint32_t *freq)
```

File: arch/risc-v/src/common/espressif/esp_mcpwm.c [hunk 10]
```diff
@@ -0,0 +1,682 @@
...
+  /* Set the clock to be used when calculating frequency */
+
+  lower->gpio_pin = pin;
+  lower->clock = group_clock / MCPWM_DEV_CLK_PRESCALE;
+
+  /* Configure GPIO pin */
+
+  esp_mcpwm_capture_set_gpio(lower);
+
+  return (struct cap_lowerhalf_s *) lower;
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/espressif/esp_mcpwm.c:678]:
```suggestion
  return (struct cap_lowerhalf_s *)lower;
```

File: boards/risc-v/esp32c6/common/src/esp_board_mcpwm.c [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+int board_capture_initialize(void)
+{
+  int ret;
+  struct cap_lowerhalf_s *cap;
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE_CH0
+  cap = esp_mcpwm_capture_initialize(0, CONFIG_ESP_MCPWM_CAPTURE_CH0_GPIO);
+
+  if (!cap)
```
Comment (pkarashchenko) [inline boards/risc-v/esp32c6/common/src/esp_board_mcpwm.c:68]:
```suggestion
  if (cap == NULL)
```

File: boards/risc-v/esp32c6/common/src/esp_board_mcpwm.c [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+  if (ret < 0)
+    {
+      syslog(LOG_ERR, "ERROR: cap_register failed: %d\n", ret);
+      return ret;
+    }
+#endif
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE_CH1
+  cap = esp_mcpwm_capture_initialize(1, CONFIG_ESP_MCPWM_CAPTURE_CH1_GPIO);
+  if (!cap)
```
Comment (pkarashchenko) [inline boards/risc-v/esp32c6/common/src/esp_board_mcpwm.c:84]:
```suggestion
  if (cap == NULL)
```

File: boards/risc-v/esp32c6/common/src/esp_board_mcpwm.c [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+ ****************************************************************************/
+
+int board_capture_initialize(void)
+{
+  int ret;
+  struct cap_lowerhalf_s *cap;
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE_CH0
+  cap = esp_mcpwm_capture_initialize(0, CONFIG_ESP_MCPWM_CAPTURE_CH0_GPIO);
+
```
Comment (pkarashchenko) [inline boards/risc-v/esp32c6/common/src/esp_board_mcpwm.c:67]:
```suggestion
```

File: boards/risc-v/esp32c6/common/src/esp_board_mcpwm.c [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+  if (ret < 0)
+    {
+      syslog(LOG_ERR, "ERROR: cap_register failed: %d\n", ret);
+      return ret;
+    }
+#endif
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE_CH2
+  cap = esp_mcpwm_capture_initialize(2, CONFIG_ESP_MCPWM_CAPTURE_CH2_GPIO);
+  if (!cap)
```
Comment (pkarashchenko) [inline boards/risc-v/esp32c6/common/src/esp_board_mcpwm.c:100]:
```suggestion
  if (cap == NULL)
```

File: boards/risc-v/esp32h2/common/src/esp_board_mcpwm.c [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+ ****************************************************************************/
+
+int board_capture_initialize(void)
+{
+  int ret;
+  struct cap_lowerhalf_s *cap;
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE_CH0
+  cap = esp_mcpwm_capture_initialize(0, CONFIG_ESP_MCPWM_CAPTURE_CH0_GPIO);
+
```
Comment (pkarashchenko) [inline boards/risc-v/esp32h2/common/src/esp_board_mcpwm.c:67]:
```suggestion
```

File: boards/risc-v/esp32h2/common/src/esp_board_mcpwm.c [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+  if (!cap)
```
Comment (pkarashchenko) [inline boards/risc-v/esp32h2/common/src/esp_board_mcpwm.c:68]:
```suggestion
  if (cap == NULL)
```

File: boards/risc-v/esp32h2/common/src/esp_board_mcpwm.c [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+  if (ret < 0)
+    {
+      syslog(LOG_ERR, "ERROR: cap_register failed: %d\n", ret);
+      return ret;
+    }
+#endif
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE_CH1
+  cap = esp_mcpwm_capture_initialize(1, CONFIG_ESP_MCPWM_CAPTURE_CH1_GPIO);
+  if (!cap)
```
Comment (pkarashchenko) [inline boards/risc-v/esp32h2/common/src/esp_board_mcpwm.c:84]:
```suggestion
  if (cap == NULL)
```

File: boards/risc-v/esp32h2/common/src/esp_board_mcpwm.c [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+  if (ret < 0)
+    {
+      syslog(LOG_ERR, "ERROR: cap_register failed: %d\n", ret);
+      return ret;
+    }
+#endif
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE_CH2
+  cap = esp_mcpwm_capture_initialize(2, CONFIG_ESP_MCPWM_CAPTURE_CH2_GPIO);
+  if (!cap)
```
Comment (pkarashchenko) [inline boards/risc-v/esp32h2/common/src/esp_board_mcpwm.c:100]:
```suggestion
  if (cap == NULL)
```

**Comment by fdcavalcanti** [comment]
@tmedicci @eren-terzioglu

File: arch/risc-v/src/common/espressif/esp_mcpwm.c (hunk 6 above)
Comment (fdcavalcanti) [reply arch/risc-v/src/common/espressif/esp_mcpwm.c:234]: Removing.

File: arch/risc-v/src/common/espressif/esp_mcpwm.c (hunk 7 above)
Comment (fdcavalcanti) [reply arch/risc-v/src/common/espressif/esp_mcpwm.c:280]: Wouldn't hurt. Just added.

**Comment by fdcavalcanti** [comment]
Applied recommended fixes on linting. Thanks for the review @pkarashchenko 
Can't the nxstyle recommend those fixes?

**Review by acassis — APPROVED** [review]

**Pull request merged**